pkg/scheduler/queue/tenant_querier_assignment.go

```diff
@@ -358,3 +311,152 @@ func (tqa *tenantQuerierAssignments) shuffleTenantQueriers(tenantID TenantID, sc
 	tqa.tenantQuerierIDs[tenantID] = querierIDSet
 	return true
 }
+
+// dequeueGetNode chooses the next node to dequeue from based on tenantIDOrder and tenantOrderIndex, which are
+// shared across all nodes to maintain an O(n) (where n = # tenants) time-to-dequeue for each tenant.
+// If tenant order were maintained by individual nodes, we would end up with O(mn) (where m = # query components)
+// time-to-dequeue for a given tenant.
+//
+// tenantOrderIndex is incremented, checks if the tenant at that index is a child of the current node (it may not be,
+// e.g., in the case that a tenant has queries queued for one query component but not others), and if so, returns
+// the tenant node if currentQuerier can handle queries for that tenant.
+//
+// Note that because we use the shared  tenantIDOrder and tenantOrderIndex to manage the queue, we functionally
+// ignore each Node's individual queueOrder and queuePosition.
+func (tqa *tenantQuerierAssignments) dequeueSelectNode(node *Node) (*Node, bool) {
```

Comment:
going to suggest some comments here I added myself locally to help me grok the control flow, take em or leave em!